packages/grafana-prometheus/src/datasource.ts

```diff
@@ -118,48 +112,106 @@ export class PrometheusDatasource
 
-    this.cache = new QueryCache({
-      getTargetSignature: this.getPrometheusTargetSignature.bind(this),
-      overlapString: instanceSettings.jsonData.incrementalQueryOverlapWindow ?? defaultPrometheusQueryOverlapWindow,
-      applyInterpolation: this.interpolateString.bind(this),
-    });
-
+    // INHERITED PROPERTIES
     this.annotations = PrometheusAnnotationSupport(this);
+    this.variables = new PrometheusVariableSupport(this, this.templateSrv);
+
+    // LANGUAGE PROVIDER
+    // This needs to be the last thing we initialize.
     this.languageProvider = languageProvider ?? new PrometheusLanguageProvider(this);
   }
 
-  init = async () => {
+  /**
+   * Initializes the Prometheus datasource by loading recording rules and checking exemplar availability.
+   *
+   * This method performs two key initialization tasks: Loads recording rules from the
+   * Prometheus API and checks if exemplars are available by testing the exemplars API endpoint.
+   */
+  init = async (): Promise<void> => {
     if (!this.disableRecordingRules) {
       this.loadRules();
     }
     this.exemplarsAvailable = await this.areExemplarsAvailable();
   };
 
+  /**
+   * Loads recording rules from the Prometheus API and extracts rule mappings.
+   *
+   * This method fetches rules from the `/api/v1/rules` endpoint and processes
+   * them to create a mapping of rule names to their corresponding queries and labels.
+   * The rules API is experimental, so errors are logged but not thrown.
+   */
+  private async loadRules(): Promise<void> {
+    try {
+      const params = {};
+      const options = { showErrorAlert: false };
+      const res = await this.metadataRequest('/api/v1/rules', params, options);
+      const ruleGroups = res.data?.data?.groups;
+
+      if (ruleGroups) {
+        this.ruleMappings = extractRuleMappingFromGroups(ruleGroups);
+      }
+    } catch (err) {
+      console.log('Rules API is experimental. Ignore next error.');
+      console.error(err);
+    }
+  }
+
+  /**
+   * Checks if exemplars are available by testing the exemplars API endpoint.
+   *
+   * This method makes a test request to the `/api/v1/query_exemplars` endpoint to determine
+   * if the Prometheus instance supports exemplars. The test uses a simple query with a
+   * 30-minute time range. If the request succeeds with a 'success' status, exemplars
+   * are considered available. Errors are caught and return false to avoid breaking
+   * the datasource initialization.
+   */
+  private async areExemplarsAvailable(): Promise<boolean> {
+    try {
+      const params = {
+        query: 'test',
+        start: dateTime().subtract(30, 'minutes').valueOf().toString(),
+        end: dateTime().valueOf().toString(),
+      };
+      const options = { showErrorAlert: false };
+      const res = await this.metadataRequest('/api/v1/query_exemplars', params, options);
+
+      return res.data.status === 'success';
+    } catch (err) {
+      return false;
+    }
+  }
+
```

Comment:
There are some external consumers of the grafana/prometheus package so making `loadRules` and `areExemplarsAvailable` private will be a breaking change. Is this an intentional breaking change?

Reply:
I didn't dig to deep into this, I did this search https://github.com/search?q=org%3Agrafana%20%40grafana%2Fprometheus&type=code and there was 5 pages of result so I just wanted you to be aware of the potential breaking change. If you have dug deeper and couldn't find any consumers I'm alright with that. Sorry for the noise 👍

Reply:
i just did a deeper dive into this, and i also didn't find any usage. i appreciate you raising this though!